x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java

```diff
@@ -0,0 +1,313 @@
+        force = parser.acceptsAll(List.of("f", "force"),
+            "Use this option to force execution of the command against a cluster that is currently unhealthy.");
+    }
+
+    @Override
+    public void close() {
+        if (password != null) {
+            password.close();
+        }
+    }
+
+    @Override
+    protected final void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        validate(terminal, options, env);
+        KeyStoreWrapper keyStoreWrapper = keyStoreFunction.apply(env);
+        final Environment newEnv;
+        final Settings settings;
+        if (keyStoreWrapper != null) {
+            decryptKeyStore(keyStoreWrapper, terminal);
+            Settings.Builder settingsBuilder = Settings.builder();
+            settingsBuilder.put(env.settings(), true);
+            if (settingsBuilder.getSecureSettings() == null) {
+                settingsBuilder.setSecureSettings(keyStoreWrapper);
+            }
+            settings = settingsBuilder.build();
+            newEnv = new Environment(settings, env.configFile());
+        } else {
+            newEnv = env;
+            settings = env.settings();
+        }
+
+        ensureFileRealmEnabled(settings);
+        try {
+            final Hasher hasher = Hasher.resolve(XPackSettings.PASSWORD_HASHING_ALGORITHM.get(settings));
+            password = new SecureString(generatePassword(PASSWORD_LENGTH));
+            username = generateUsername();
+
+            final Path passwordFile = FileUserPasswdStore.resolveFile(newEnv);
+            final Path rolesFile = FileUserRolesStore.resolveFile(newEnv);
+            FileAttributesChecker attributesChecker = new FileAttributesChecker(passwordFile, rolesFile);
+            // Store the roles file first so that when we get to store the user, it will definitely be a superuser
+            Map<String, String[]> userRoles = FileUserRolesStore.parseFile(rolesFile, null);
+            if (userRoles == null) {
+                throw new IllegalStateException("File realm configuration file [" + rolesFile + "] is missing");
+            }
+            userRoles = new HashMap<>(userRoles);
+            userRoles.put(username, ROLES);
+            FileUserRolesStore.writeFile(userRoles, rolesFile);
+
+            Map<String, char[]> users = FileUserPasswdStore.parseFile(passwordFile, null, settings);
+            if (users == null) {
+                throw new IllegalStateException("File realm configuration file [" + passwordFile + "] is missing");
+            }
+            users = new HashMap<>(users);
+            users.put(username, hasher.hash(password));
+            FileUserPasswdStore.writeFile(users, passwordFile);
+
+            attributesChecker.check(terminal);
+            final boolean forceExecution = options.has(force);
+            checkClusterHealthWithRetries(newEnv, terminal, 5, forceExecution);
+            executeCommand(terminal, options, newEnv);
```

Comment:
nit: I would pass the username and password to this method instead of making them class members. It's more clear the lifecycle like this.

Reply:
we'd need to pass them as parameters to all checkClusterHealthWithRetries, executeCommand, cleanup. I like the fact that we keep ownership of the password here and close it in all cases etc. Happy to change, but I miss the point of "It's more clear the lifecycle like this." , I'll ping you to chat

Reply:
we chatted and I made the change as suggested